### ISSUE TITLE
open_ai_gym_async error

### ISSUE THREAD

**Issue by Gor-Ren** [reinforceio/tensorforce#432]
Thanks for the great package. I can successfully run single agents using the example code and am now trying to run the distributed [open_ai_gym_async](https://github.com/reinforceio/tensorforce/blob/master/examples/openai_gym_async.py) example on the CartPole environment, using a command like the one suggested in the file:

`python examples/openai_gym_async.py CartPole-v0 -a examples/configs/vpg.json -n examples/configs/mlp2_network.json -e 10000 -m 200 -W 3`

The script creates tmux sessions as intended and I see four Python processes start for a couple of seconds, but then they terminate.

To investigate why, I run open_ai_gym_async with the command line arg generated for the parameter_server process (these are normally sent to the param. server tmux session): `CartPole-v0 -a examples/configs/vpg.json -n examples/configs/mlp2_network.json --num-workers 3 --child --task-index 0 --episodes 10000 --max-episode-timesteps 200 --parameter-server`

This throws an exception with the following stack trace:

```Traceback (most recent call last):
  File "/home/gordon/dev/aircraft-rl/tensorforce_examples/openai_gym_async.py", line 243, in <module>
    main()
  File "/home/gordon/dev/aircraft-rl/tensorforce_examples/openai_gym_async.py", line 200, in main
    network=network
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/agents/agent.py", line 283, in from_spec
    kwargs=kwargs
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/util.py", line 192, in get_object
    return obj(*args, **kwargs)
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/agents/vpg_agent.py", line 136, in __init__
    entropy_regularization=entropy_regularization
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/agents/learning_agent.py", line 141, in __init__
    batching_capacity=batching_capacity
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/agents/agent.py", line 80, in __init__
    self.model = self.initialize_model()
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/agents/vpg_agent.py", line 163, in initialize_model
    gae_lambda=self.gae_lambda
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/pg_model.py", line 98, in __init__
    requires_deterministic=False
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/distribution_model.py", line 90, in __init__
    discount=discount
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/memory_model.py", line 114, in __init__
    reward_preprocessing=reward_preprocessing
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/model.py", line 221, in __init__
    self.setup()
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/model.py", line 296, in setup
    self.setup_saver()
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/model.py", line 545, in setup_saver
    global_variables = self.global_model.get_variables(include_submodules=True, include_nontrainable=True)
AttributeError: 'NoneType' object has no attribute 'get_variables'
```

I walked through the control flow in the debugger and the problem seems to occur when running `agent = Agent.from_spec(...)` for the parameter server. During initialisation, I _think_ this parameter server is supposed to start a server and block. Instead, it begins trying to setup a model and eventually tries to invoke a method on an uninitialised `self.global_model`. The problem relates to [Model](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/models/model.py)`.setup()` on line 235:

        if self.execution_type == "distributed" and not self.server and not self.is_local_model:
            self.start_server()

This code block will never run for the parameter server because `self.is_local_model` is initialised True. I tried a fix:

        if (self.execution_type == "distributed" and not self.server and not self.is_local_model) \
                 or self.distributed_spec["job"] == "ps":
             self.start_server()

This correctly starts the grpc server, confirmed by TF console outputs, and blocks there. Unfortunately the overall behaviour is unchanged: although the ps now starts correctly, when a worker is started in another process a different exception is thrown:

```
Traceback (most recent call last):
  File "/home/gordon/dev/aircraft-rl/tensorforce_examples/openai_gym_async.py", line 243, in <module>
    main()
  File "/home/gordon/dev/aircraft-rl/tensorforce_examples/openai_gym_async.py", line 200, in main
    network=network
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/agents/agent.py", line 283, in from_spec
    kwargs=kwargs
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/util.py", line 192, in get_object
    return obj(*args, **kwargs)
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/agents/vpg_agent.py", line 136, in __init__
    entropy_regularization=entropy_regularization
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/agents/learning_agent.py", line 141, in __init__
    batching_capacity=batching_capacity
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/agents/agent.py", line 80, in __init__
    self.model = self.initialize_model()
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/agents/vpg_agent.py", line 163, in initialize_model
    gae_lambda=self.gae_lambda
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/pg_model.py", line 98, in __init__
    requires_deterministic=False
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/distribution_model.py", line 90, in __init__
    discount=discount
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/memory_model.py", line 114, in __init__
    reward_preprocessing=reward_preprocessing
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/model.py", line 221, in __init__
    self.setup()
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/model.py", line 309, in setup
    self.setup_session(self.server, hooks, graph_default_context)
  File "/home/gordon/dev/aircraft-rl/venv/lib/python3.6/site-packages/tensorforce/models/model.py", line 718, in setup_session
    master=server.target,
AttributeError: 'NoneType' object has no attribute 'target'
```
The problem here is that the logic in Model.setup() does not init `self.server` if the Model is a local model, but then tries to access `self.server.target`. 

I can infer from comments in the code (but I don't really know) that a local model should have a server, and a global model should not. Then we can instead replace line 235 of [model.py](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/models/model.py) to:

        if self.execution_type == "distributed" and not self.server and self.is_local_model:
            self.start_server()
(N.B. `and not self.is_local_model` -> `and self.is_local_model`)

With this fix I can run the script and happily tmux into the session. Hurrah! Unfortunately no work actually seems to be happening: there are no Python sessions listed using `top` and each of the worker tmux windows only display some (apparently harmless) TF warnings/errors:

```
││ 2018-06-12 15:51:24.056491: W tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc:1441] Failed to build SimpleGraphView. 
││ 2018-06-12 15:51:24.057334: E tensorflow/core/grappler/optimizers/dependency_optimizer.cc:586] Iteration = 0, topological sort failed with message: Non-existent input ^ConstantFoldingCt 
││ g-worker2/cond/add/Switch_0 for node ConstantFolding/vpg-worker2/cond/strided_slice_1/stack_const_axis                                                                                    
││ 2018-06-12 15:51:24.057729: E tensorflow/core/grappler/optimizers/dependency_optimizer.cc:586] Iteration = 1, topological sort failed with message: Non-existent input ^ConstantFoldingCt 
││ g-worker2/cond/add/Switch_0 for node ConstantFolding/vpg-worker2/cond/strided_slice_2/stack_const_axis
```
I say these are apparently harmless because I receive the exact same messages when running the single (non-distributed) OpenAI example script, and it still happily runs and learns without issue.

I have rather reached the end of my ability to debug this issue, can anyone assist?

I am running Ubuntu 18.04, Python 3.6, TF1.8 (cpu), TensorForce 4.2. Files are being run from PyCharm.

**Comment by Gor-Ren** [comment]
Update: I resolved the final problem I mentioned where the worker processes hang and don't do anything. From a hunch based on the tensorflow-gpu issues for version >1.5 (issue #391) I rolled back from tensorflow (cpu) 1.8 to 1.5.1 on a hunch and now have learning. (This is still using the modification of model.py line 235 `and not self.is_local_model` -> `and self.is_local_model` described above.)

Some improvements could still be made to the openai_gym_async.py, for example logging doesn't seem to be working, which I will try and look into.

Perhaps an author more familiar with the intent of the Model code could confirm whether the described modification of model.py line 235 is correct and get a fix pushed. I'd also consider updating the README to recommend TensorFlow v 1.5.1 for CPU and GPU given these issues.
